src/compiler/builder.ts

```diff
@@ -1038,8 +1076,30 @@ namespace ts {
         return { host, newProgram, oldProgram, configFileParsingDiagnostics: configFileParsingDiagnostics || emptyArray };
     }
 
-    export function computeSignature(text: string, data: WriteFileCallbackData | undefined, computeHash: BuilderState.ComputeHash | undefined) {
-        return BuilderState.computeSignature(data?.sourceMapUrlPos !== undefined ? text.substring(0, data.sourceMapUrlPos) : text, computeHash);
+    function getTextHandlingSourceMapForSignature(text: string, data: WriteFileCallbackData | undefined) {
+        return data?.sourceMapUrlPos !== undefined ? text.substring(0, data.sourceMapUrlPos) : text;
+    }
+
+    export function computeSignatureWithDiagnostics(
+        program: Program,
+        text: string,
+        computeHash: BuilderState.ComputeHash | undefined,
+        getCanonicalFileName: GetCanonicalFileName,
+        data: WriteFileCallbackData | undefined
+    ) {
+        text = getTextHandlingSourceMapForSignature(text, data);
+        if (data?.diagnostics?.length) {
+            text += formatDiagnostics(data.diagnostics, {

```

Comment:
I don't believe anyone will ever see this string, correct?  If so, can we skip some work?  (Pretty printing, indenting, typeNodeToTypeString, etc)

Reply:
Can `getCanonicalFileName` be the identity function?

Reply:
it isnt doing much pretty printing apart from indenting multi line errors but i can get rid of that. 
These diagnostics were already being generated.. we are only using them so `TypeNodeToTypeString` cost isnt extra in this PR. I believe we should retain those for correctness.